hedera-node/hedera-app/src/main/java/com/hedera/node/app/roster/schemas/V0540RosterSchema.java

```diff
@@ -0,0 +1,78 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.roster.schemas;
+
+import com.hedera.hapi.node.base.SemanticVersion;
+import com.hedera.hapi.node.state.primitives.ProtoBytes;
+import com.hedera.hapi.node.state.roster.Roster;
+import com.hedera.hapi.node.state.roster.RosterState;
+import com.swirlds.state.spi.MigrationContext;
+import com.swirlds.state.spi.Schema;
+import com.swirlds.state.spi.StateDefinition;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Roster Schema
+ */
+public class V0540RosterSchema extends Schema {
+    private static final Logger log = LogManager.getLogger(V0540RosterSchema.class);
+    public static final String ROSTER_KEY = "ROSTER";
+    public static final String ROSTER_STATES_KEY = "ROSTER_STATE";
+    /** this can't be increased later so we pick some number large enough, 2^16. */
+    private static final long MAX_ROSTERS = 65_536L;
+
+    /**
+     * The version of the schema.
+     */
+    private static final SemanticVersion VERSION =
+            SemanticVersion.newBuilder().major(0).minor(54).patch(0).build();
+
+    /**
+     * Create a new instance
+     */
+    public V0540RosterSchema() {
+        super(VERSION);
+    }
+
+    @NonNull
+    @Override
+    public Set<StateDefinition> statesToCreate() {
+        return Set.of(
+                StateDefinition.singleton(ROSTER_STATES_KEY, RosterState.PROTOBUF),
+                StateDefinition.onDisk(ROSTER_KEY, ProtoBytes.PROTOBUF, Roster.PROTOBUF, MAX_ROSTERS));
+    }
+
+    @Override
+    public void migrate(@NonNull final MigrationContext ctx) {
+        if (ctx.previousVersion() == null) {
+            log.info("Creating genesis roster and roster state");
+            // This migration code is really, for now, a default value provider.
+            // At genesis we put empty roster and roster state into the state
+            // as serialization/deserialization fails on null.
+            // It is important to distinguish this from the Genesis Roster,
+            // which will be provided by the Execution layer depending on
+            // the desired network mode
+            final var rosterState = ctx.newStates().getSingleton(ROSTER_STATES_KEY);
+            rosterState.put(RosterState.DEFAULT);
+            final var roster = ctx.newStates().get(ROSTER_KEY);
+            roster.put(ProtoBytes.DEFAULT, Roster.DEFAULT);
```

Comment:
I'm unsure about the line 75. It populates the map with a random entry that is never going to be removed (unless we implement a special logic somewhere, which is ugly.)

Can the map be initially empty? I.e. have no entries at all. If that's impossible, then my suggestion would be to populate it with the actual current roster as found in the PlatformState (i.e. the AddressBook converted to Roster.)

In fact, same applies to the RosterState just above. I think it should point to the actual active roster from the beginning and have a corresponding entry in the list of pairs<round, roster_hash>.

It would be difficult and unintuitive to support the initialization of these data structures somewhere else. I believe this initial initialization belongs here.

Reply:
on revisiting, it makes sense to have an initial empty map, but i wasn't sure of the state behaviour. I just tried with a local test and it shows it's ok to do, so i've removed it. Thank you.

But the rosterstate has to stay, as it has to be initialized. It will be subsequently initialized here depending on the network mode configuration, but the flow to arrive at that configuration is not yet done. For now, a default value provider is sufficient.